src/dev/lupio/lupio_tmr.cc

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_tmr.hh"
+
+#include "cpu/base.hh"
+#include "debug/LupioTMR.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioTMR.hh"
+
+// Specific fields for CTRL
+#define LUPIO_TMR_IE    0x1
+#define LUPIO_TMR_PD    0x2
+
+// Specific fields for STAT
+#define LUPIO_TMR_EX    0x1
+
+namespace gem5
+{
+using namespace RiscvISA;
+LupioTMR::LupioTMR(const Params &params) :
+    BasicPioDevice(params, params.pio_size),
+    system(params.system),
+    tmrEvent([this]{ lupioTMRCallback(); }, name()) 
+{
+    start = curTick(); 
+    DPRINTF(LupioTMR, "LupioTMR initalized\n");
+}
+
+void LupioTMR::updateIRQ(int level)
+{
+    auto tc = system->threads[0];
```

Comment:
I don't think you can always assume that system->threads[0] is what you want. Let's talk about a better solution in our meeting.